### DISCUSSION TITLE
Finalizers and Memory Usage

### DISCUSSION THREAD

**Discussion by Larhei** [dotnet/maui#12026] [Q&A] [answered]
Hi,

because i´m not sure if this is an excepted behavior I wanted to first ask before opening a bug for this.
In the dot net world we are used to have finalizers, that are called by the GC when cleaning up.
In Maui have the problem that none of my finalizers are called on iOS and some are not called on Android and Windows.
So I would say we have a few memory leaks... But because of missing Profilers I  can not say for sure. So could someone please give a hint if finalizers should be called or not?
Could someone please also give a hint how to memory profile under iOS and Android?

**Comment by Xyncgas** [comment]
I remember the trouble it did me when my objects aren't disposed because GC just doesn't care.

I don't know if this is the case for you or is GC bugged, but..

[ .NET MAUI Shell lifecycle ], through this you can manually do any 'finalizing' you want. You can also call ```GC.Collect``` and see if it cares

**Reply by Larhei** [reply]
hi @Xyncgas,

thanks for the advice. But OnDisapering does not indicate if page was poped from backstack or not. I agree freeing resources in OnNavigatedFrom or OnDisapering helps. But at the same time it does not help in case of navgating back to the page after pushing an other view onto the stack. 

And of course I could some navigation service in order to keep track if the view was poped of the stack during OnDisapering. 

But I really try to understand if or if not Finalizers should be called by GC on all platforms. 

And if not... What would be a indicator that our views where collected. On iOs there is didReceiveMemoryWarning. Is there anything like that in maui infrastructure? 
And is [this](https://learn.microsoft.com/en-us/xamarin/ios/internals/api-design/#Memory_Management) still valid in maui...

**Comment by hartez** [comment, marked as answer]
There's no special reason in MAUI that finalizers would not run, assuming the object has met all the requirements for being finalized. It's possible that what you are seeing are memory leaks. I would suggest creating a minimal repro project which demonstrates your finalizer not being called, and post that in an issue. Then we can take a look and see if we have a memory leak with that particular component.

**Reply by Larhei** [reply]
Thanks for the reply. Repro is [here](https://github.com/dotnet/maui/issues/12048).
